File: csharp-dotnet/OnetWebService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net.Http.Headers;
using Newtonsoft.Json.Linq;

public class OnetWebService
{
    public class QueryParams : IEnumerable<KeyValuePair<string, string>>
    {
        private List<KeyValuePair<string, string>> internalParams = new List<KeyValuePair<string, string>>();
        public IEnumerator<KeyValuePair<string, string>> GetEnumerator() => internalParams.GetEnumerator();
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => internalParams.GetEnumerator();

        public void Add(string key, string value) => internalParams.Add(new KeyValuePair<string, string>(key, value));
    }

    private string baseURL;
    private HttpClient client;

    private static string EncodeAuth(string username, string password)
    {
        return Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(username + ":" + password));
    }

    public OnetWebService(string username, string password)
    {
        HttpClientHandler handler = new HttpClientHandler()
        {
            AllowAutoRedirect = false,
        };
        client = new HttpClient(handler);
        client.Timeout = new TimeSpan(0, 0, 10);
        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        client.DefaultRequestHeaders.UserAgent.ParseAdd("dotnet-OnetWebService/1.00 (bot)");
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", EncodeAuth(username, password));

        SetVersion();
    }

    public void SetVersion(string version = "")
    {
        if (version == "")
        {
            baseURL = "https://services.onetcenter.org/ws/";
        }
        else
        {
            baseURL = "https://services.onetcenter.org/ws/v" + version + "/ws/";
        }
    }

    public async Task<JObject> Call(string path, QueryParams query = null)
    {
        List<string> encoded_params = new List<string>();
        if (query != null)
        {
            foreach (KeyValuePair<string, string> pair in query)
            {
                encoded_params.Add(System.Net.WebUtility.UrlEncode(pair.Key) + "=" + System.Net.WebUtility.UrlEncode(pair.Value));
            }
        }
        string url = baseURL + path;
        if (encoded_params.Count > 0)
        {
            url += "?" + String.Join("&", encoded_params.ToArray());
        }

        JObject result = new JObject();
        result["error"] = new JValue("Call to " + url + " failed with unknown reason");
        try
        {
            HttpResponseMessage response = await client.GetAsync(url);
            if (response.StatusCode == (System.Net.HttpStatusCode)200 || response.StatusCode == (System.Net.HttpStatusCode)422)
            {
                result = JObject.Parse(await response.Content.ReadAsStringAsync());
            }
            else
            {
                result["error"] = new JValue("Call to " + url + " failed with error code " + ((int)response.StatusCode).ToString());
            }
        }
        catch (HttpRequestException e)
        {
            result["error"] = new JValue("Call to " + url + " failed with reason: " + e.Message);
        }

        return result;
    }
}

File: python-3/OnetWebService.py
import urllib.request, urllib.parse, urllib.error
import urllib.request, urllib.error, urllib.parse
import base64
import json

class OnetWebService:
    
    def __init__(self, username, password):
        self._headers = {
            'User-Agent': 'python-OnetWebService/1.00 (bot)',
            'Authorization': 'Basic ' + base64.standard_b64encode((username + ':' + password).encode()).decode(),
            'Accept': 'application/json' }
        self.set_version()
    
    def set_version(self, version = None):
        if version is None:
            self._url_root = 'https://services.onetcenter.org/ws/'
        else:
            self._url_root = 'https://services.onetcenter.org/v' + version + '/ws/'
    
    def call(self, path, *query):
        url = self._url_root + path
        if len(query) > 0:
            url += '?' + urllib.parse.urlencode(query, True)
        req = urllib.request.Request(url, None, self._headers)
        handle = None
        try:
            handle = urllib.request.urlopen(req)
        except urllib.error.HTTPError as e:
            if e.code == 422:
                return json.load(e)
            else:
                return { 'error': 'Call to ' + url + ' failed with error code ' + str(e.code) }
        except urllib.error.URLError as e:
            return { 'error': 'Call to ' + url + ' failed with reason: ' + str(e.reason) }
        code = handle.getcode()
        if (code != 200) and (code != 422):
            return { 'error': 'Call to ' + url + ' failed with error code ' + str(code),
                     'urllib2_info': handle }
        return json.load(handle)
